src/NuGet.Clients/NuGet.CommandLine/Common/SelfUpdater.cs

```diff
@@ -59,96 +58,95 @@ public string AssemblyLocation
-                Console.WriteLine(LocalizedResourceManager.GetString("UpdateCommandUpdatingNuGet"), package.Version);
+                _console.WriteLine(LocalizedResourceManager.GetString("UpdateCommandUpdatingNuGet"), latestVersion);
 
-                // Get NuGet.exe file from the package
-                CoreV2.NuGet.IPackageFile file = package.GetFiles().FirstOrDefault(f => Path.GetFileName(f.Path).Equals(NuGetExe, StringComparison.OrdinalIgnoreCase));
+                var packageIdentity = new PackageIdentity(NuGetCommandLinePackageId, latestVersion);
 
-                // If for some reason this package doesn't have NuGet.exe then we don't want to use it
-                if (file == null)
+                var tempDir = Path.Combine(NuGetEnvironment.GetFolderPath(NuGetFolderPath.Temp), "updateSelf");
+                var nupkgPath = FileUtility.GetTempFilePath(tempDir);
+                try
                 {
-                    throw new CommandLineException(LocalizedResourceManager.GetString("UpdateCommandUnableToLocateNuGetExe"));
+                    DirectoryUtility.CreateSharedDirectory(tempDir);
+
+                    var downloadResourceResult = await PackageDownloader.GetDownloadResourceResultAsync(
+                                        sourceRepository,
+                                        packageIdentity,
+                                        new PackageDownloadContext(sourceCacheContext),
+                                        tempDir,
+                                        _console,
+                                        cancellationToken);
+
+                    // Get the exe path and move it to a temp file (NuGet.exe.old) so we can replace the running exe with the bits we got 
+                    // from the package repository
+                    var nugetExeFile = (await downloadResourceResult.PackageReader.GetFilesAsync(CancellationToken.None)).FirstOrDefault(f => Path.GetFileName(f).Equals(NuGetExe, StringComparison.OrdinalIgnoreCase));
+
+                    // If for some reason this package doesn't have NuGet.exe then we don't want to use it
+                    if (nugetExeFile == null)
+                    {
+                        throw new CommandLineException(LocalizedResourceManager.GetString("UpdateCommandUnableToLocateNuGetExe"));
+                    }
+
+                    string renamedPath = exePath + ".old";
+
+                    FileUtility.Move(exePath, renamedPath);
+
+                    using (Stream fromStream = await downloadResourceResult.PackageReader.GetStreamAsync(nugetExeFile, cancellationToken), toStream = File.Create(exePath))
+                    {
+                        fromStream.CopyTo(toStream);
+                    }
+                }
+                finally
+                {
+                    // Delete the temporary directory
+                    try
+                    {
+                        Directory.Delete(tempDir, recursive: true);
```

Comment:
Should we also delete `renamedPath`, or is the expectation that `nuget.exe.old` just stays around after update?

Reply:
It's deleted the next time the nuget.exe starts up.

https://github.com/NuGet/NuGet.Client/blob/870841c87ad76468e1712a20e51d558fc0026201/src/NuGet.Clients/NuGet.CommandLine/Program.cs#L238

That's cause you can't delete your executable.